src/renderers/dom/client/ReactEventListener.js

```diff
@@ -60,29 +62,59 @@ PooledClass.addPoolingTo(
 );
 
 function handleTopLevelImpl(bookKeeping) {
-  var topLevelTarget = ReactMount.getFirstReactDOM(
-    getEventTarget(bookKeeping.nativeEvent)
-  ) || window;
-
-  // Loop through the hierarchy, in case there's any nested components.
-  // It's important that we build the array of ancestors before calling any
-  // event handlers, because event handlers can modify the DOM, leading to
-  // inconsistencies with ReactMount's node cache. See #1105.
-  var ancestor = topLevelTarget;
-  while (ancestor) {
-    bookKeeping.ancestors.push(ancestor);
-    ancestor = findParent(ancestor);
-  }
+  var path = bookKeeping.nativeEvent.path;
+  var topLevelTargetID;
+  if (path != null) {
+    var oldRootId = undefined;
+    var currentNativeTarget = undefined;
+    for (var j = 0; j < path.length; j++) {
+      if (currentNativeTarget === undefined) {
+        currentNativeTarget = path[j];
+      }
+      if (path[j].nodeType === DOCUMENT_FRAGMENT_NODE_TYPE) {
+        currentNativeTarget = undefined;
+      }
+      var newRootId = ReactInstanceHandles.getReactRootIDFromNodeID(ReactMount.getID(path[j]));
```

Comment:
This is a relatively expensive lookup to do for every item and could be worse if we use a different ID strategy that doesn't concat. Another strategy would be to create an inner loop, after this, that keeps going until the node ID equals this root ID. That way you don't have to extract the root ID from every intermediate node.